src/analysis/normed_space/SemiNormedGroup.lean

```diff
@@ -0,0 +1,127 @@
+/-
+Copyright (c) 2021 Johan Commelin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Johan Commelin, Riccardo Brasca
+-/
+import analysis.normed_space.normed_group_hom
+import category_theory.concrete_category.bundled_hom
+import category_theory.limits.shapes.zero
+
+/-!
+# The category of seminormed groups
+
+We define `SemiNormedGroup`, the category of seminormed groups and normed group homs between them,
+as well as `SemiNormedGroup₁`, the subcategory of norm non-increasing morphisms.
+-/
+
+noncomputable theory
+
+universes u
+
+open category_theory
+
+/-- The category of seminormed abelian groups and bounded group homomorphisms. -/
+def SemiNormedGroup : Type (u+1) := bundled semi_normed_group
+
+namespace SemiNormedGroup
+
+instance bundled_hom : bundled_hom @normed_group_hom :=
+⟨@normed_group_hom.to_fun, @normed_group_hom.id, @normed_group_hom.comp, @normed_group_hom.coe_inj⟩
+
+attribute [derive [has_coe_to_sort, large_category, concrete_category]] SemiNormedGroup
+
+/-- Construct a bundled `SemiNormedGroup` from the underlying type and typeclass. -/
+def of (M : Type u) [semi_normed_group M] : SemiNormedGroup := bundled.of M
+
+instance (M : SemiNormedGroup) : semi_normed_group M := M.str
+
+@[simp] lemma coe_of (V : Type u) [semi_normed_group V] : (SemiNormedGroup.of V : Type u) = V := rfl
+@[simp] lemma coe_id (V : SemiNormedGroup) : ⇑(𝟙 V) = id := rfl
+
+instance : has_zero SemiNormedGroup := ⟨of punit⟩
+instance : inhabited SemiNormedGroup := ⟨0⟩
+
+instance : limits.has_zero_morphisms.{u (u+1)} SemiNormedGroup := {}
+
+end SemiNormedGroup
+
+/--
+`SemiNormedGroup₁` is a type synonym for `SemiNormedGroup`,
+which we shall equip with the category structure consisting only of the norm non-increasing maps.
+-/
+@[derive has_coe_to_sort]
+def SemiNormedGroup₁ : Type (u+1) := bundled semi_normed_group
+
+namespace SemiNormedGroup₁
+
+instance : large_category.{u} SemiNormedGroup₁ :=
+{ hom := λ X Y, { f : normed_group_hom X Y // f.norm_noninc },
+  id := λ X, ⟨normed_group_hom.id, normed_group_hom.norm_noninc.id⟩,
+  comp := λ X Y Z f g, ⟨g.1.comp f.1, g.2.comp f.2⟩, }
```

Comment:
Should we try to state `g.1.comp f.1` using the coercion, so that `val_coe` (below) doesn't have to be used so often?

Reply:
Done.